src/content/components/FlameChartCanvas.js

```diff
@@ -0,0 +1,190 @@
+import React, { Component, PropTypes } from 'react';
+import shallowCompare from 'react-addons-shallow-compare';
+import { timeCode } from '../../common/time-code';
+import TextMeasurement from '../../common/text-measurement';
+
+const ROW_HEIGHT = 16;
+const TEXT_OFFSET_START = 3;
+const TEXT_OFFSET_TOP = 11;
+
+class FlameChartCanvas extends Component {
+
+  constructor(props) {
+    super(props);
+    this._requestedAnimationFrame = false;
+    this._devicePixelRatio = 1;
+    this._textWidthsCache = {};
+  }
+
+  _scheduleDraw() {
+    if (!this._requestedAnimationFrame) {
+      this._requestedAnimationFrame = true;
+      window.requestAnimationFrame(() => {
+        this._requestedAnimationFrame = false;
+        if (this.refs.canvas) {
+          timeCode('FlameChartCanvas render', () => {
+            this.drawCanvas();
+          });
+        }
+      });
+    }
+  }
+
+  shouldComponentUpdate(nextProps, nextState) {
+    return shallowCompare(this, nextProps, nextState);
+  }
+
+  componentDidMount() {
+    this._textMeasurement = new TextMeasurement(this.refs.canvas.getContext('2d'));
+  }
+
+  _prepCanvas() {
+    const {canvas} = this.refs;
+    const {containerWidth, containerHeight} = this.props;
+    const {devicePixelRatio} = window;
+    const pixelWidth = containerWidth * devicePixelRatio;
+    const pixelHeight = containerHeight * devicePixelRatio;
+    if (!this._ctx) {
+      this._ctx = canvas.getContext('2d');
+    }
+    if (canvas.width !== pixelWidth || canvas.height !== pixelHeight) {
+      canvas.width = pixelWidth;
+      canvas.height = pixelHeight;
+      this._ctx.scale(this._devicePixelRatio, this._devicePixelRatio);
+    }
+    if (this._devicePixelRatio !== devicePixelRatio) {
+      // Make sure and multiply by the inverse of the previous ratio, as the scaling
+      // operates off of the previous set scale.
+      const scale = (1 / this._devicePixelRatio) * devicePixelRatio;
```

Comment:
It would probably be simpler to just throw away the whole canvas contents away in this case, by folding the "device pixel ratio changed" condition into the previous if block.